library/core/src/iter/traits/iterator.rs

```diff
@@ -48,13 +48,24 @@ pub trait Iterator {
     type Item;
 
     /// Advances the iterator and returns the next value.
+    /// Returns [`None`] when iteration is finished.
     ///
-    /// Returns [`None`] when iteration is finished. Individual iterator
-    /// implementations may choose to resume iteration, and so calling `next()`
-    /// again may or may not eventually start returning [`Some(Item)`] again at some
-    /// point.
+    /// # Behavior after [`None`] is returned
+    ///
+    /// Calling `next()` on an iterator after it has returned `None`
+    /// is in general a perfectly valid use of the interface.
+    /// Implementations may choose to resume iteration, and so calling `next()`
+    /// again may eventually start returning [`Some(Item)`] again at some
+    /// point. Alternately, implementations may choose to continue returning [`None`]
+    /// forever, in which case they should implement [`FusedIterator`] as well.
+    ///
+    /// Implementations should document their behavior, or their lack of a guarantee
+    /// about their behavior, after `None` is returned. If they must panic,
+    /// block forever, or otherwise cause problems, then this should be prominently
+    /// indicated in their docs (no different from any other potentially-misbehaving API).
```

Comment:
Would you be open to following up with such documentation on all the standard library iterators? It would help for the standard library to set a good example of what we're saying other libraries "should" do, and it would help inform this documentation here by revealing what the implications of it look like in practice. Like what it looks like when a generic iterator is "conditionally fused" (e.g. Peekable).

It can be in this PR, or a separate draft PR to merge after this one.